### PR TITLE
Fix creating channel when ${PWD} contains space.

### PR DESCRIPTION
**Pull request by WaleedMortaja**
If the path of the project contains spaces, the `test-network/scripts/envVar.sh` script sets the value of `$ORDERER_CA` to a value containg `${PWD}` which, in turn, contains space(s).

When the variable used in `test-network/scripts/createChannel.sh`, The first part of the value (before the first space) is handled as the whole value for `--cafile`. Other parts are considered to be part of the command!

I tried putting (escaped) quotes in the `test-network/scripts/envVar.sh` definition for the variable `$ORDERER_CA` to make the fix more general, but the quotation marks were sometime interpreted to be part of the path that consisted of concatenated parts somewhere and it did not work.

While this edit will fix this issue, I belive this is just a work around. I expect that there is a better way to solve the root cause of the problem instead of just fixing it in one place. Moreover, All variables/paths that may include spaces should be properly handled as well.

Thanks

### CODE REVIEW

**Review by lehors — COMMENTED** [review]
This is actually the right way to handle the possibility of having spaces in the variable value but this should be done throughout. Do you care to submit a more comprehensive fix?
Thanks.

**Comment by lehors** [comment]
Oh, and you need to add a sign-off to your commit log for this to be mergeable.

**Comment by WaleedMortaja** [comment]
> This is actually the right way to handle the possibility of having spaces in the variable value but this should be done throughout. Do you care to submit a more comprehensive fix?
> Thanks.

I am still new to hyperledger fabric, but yes I will try to fix all other variables and make a commit.

> Oh, and you need to add a sign-off to your commit log for this to be mergeable.

Sorry for that, I will fix it also.

**Comment by nikhil550** [comment]
If we are going to fix everywhere, should that fix be applied to the envVar.sh file? (to reduce the clutter elsewhere)

**Comment by WaleedMortaja** [comment]
@nikhil550 
> If we are going to fix everywhere, should that fix be applied to the envVar.sh file? (to reduce the clutter elsewhere)

I already tried changing:
`export ORDERER_CA=${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem`

to
`export ORDERER_CA=\"${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem\"`

but I got weired erors: path(s) that contained a double-quotes in the middle of their content. The path was a `config` folder which made me wonder how the double-quote appeared there.

If you have a better idea to fix the `envVar.sh`, please suggest it.
In the meanwhile I am working the hardway hhh.

**Comment by lehors** [comment]
Indeed, this has no chance to work. The quote becomes part of the value being stored.
There is no way around it, we need to quote the variable at the time it is evaluated so that the command being called sees it as one single string, no matter whether it contains whitespaces.

**Comment by WaleedMortaja** [comment]
Hello @lehors 
Can you help me pushing the commits?
I cloned this reposotory locally, checkedout this repo (`gh pr checkout 402`), amended the commit and signed it off (`git commit --amend --signoff`), made alot of other changes, and now I am trying to push it, but it does not work. it seems that I am pushing to this repo which I do not have write access to. I tried changing the remote of my local repo to be my forked repo but it did not work too. I used the command `git push --force-with-lease myrepo patch-1` but it did not work also.

**Comment by WaleedMortaja** [comment]
> Hello @lehors
> Can you help me pushing the commits?
> I cloned this repository locally, checked out this repo (`gh pr checkout 402`), amended the commit and signed it off (`git commit --amend --signoff`), made a lot of other changes, and now I am trying to push it, but it does not work. it seems that I am pushing to this repo which I do not have write access to. I tried changing the remote of my local repo to be my forked repo but it did not work too. I used the command `git push --force-with-lease myrepo patch-1` but it did not work also.

[ This comment is just informative for steps taken to amend the commit and re-push ]
TLDR; For your local clone, add your forked repo as a remote repo, fetch/pull from it to update the references, then push/force push to it.

I figured it out. According to DCO bot, I had to sign my commit, so I had to amend my commit, sign it, force push it.
According to the DCO bot I could checkout the pull request locally using `gh pr checkout 402`.
So, 
1. First I got a fresh clone of **this** repo `git clone https://github.com/hyperledger/fabric-samples.git`.
2. Checked out the pull request  `gh pr checkout 402`.
3. Amended and signed the commit `git commit --amend --signoff`.
4. Made more changes locally and committed them.
5. In the mean time, a new commit was pushed to the master here, that is not in my local repo.
6. To get the commit locally and rebase `git rebase origin/master`.
7. To push the local commits, you have to push them to your forked repo, so first add it as a remote repo `git remote add myrepo https://github.com/WaleedMortaja/fabric-samples`.
8. Trying to push to `myrepo` failed. The reason was that I never fetched or updated its references, so `git fetch myrepo master`
9. Update my forked master `git checkout master` and then `git push myrepo master`.
10. Return to this branch `git checkout patch-1` (The local branch has two more commits).
11. Push the local commits with force to override the non-signed commit `git push myrepo patch-1 --force-with-lease`

**Review by denyeart — DISMISSED** [review]
It looks good to me but it sounds like @lehors is the expert so I'll defer to his review before merging.

**Comment by WaleedMortaja** [comment]
> It looks good to me but it sounds like @lehors is the expert so I'll defer to his review before merging.

@denyeart Thank you.
Please note that this commit is **not** 100% covering. I would recommend adding more commits and looking at the [commit message](https://github.com/hyperledger/fabric-samples/pull/402/commits/7ab304f5f5d419b7b73f303124390d3eb3e06165) before merging.
I may amend the last commit if I could fix the remaning known variable (`TLSINFO` and/or `PEER_CONN_PARMS`). Note: The commit message seems to be refereing a wrong variable here.

**Comment by lehors** [comment]
Indeed, while this definitely moves us forward (Thanks!), it is not enough. And unfortunately there is a piece of code that will be tricky to fix: parsePeerConnectionParameters in envVar.sh. I will see if I can find a way but keeping the necessary quotes through the multiple levels of evaluation is not trivial.

**Comment by WaleedMortaja** [comment]
@lehors 
I found a solution. Have a look [here](https://github.com/WaleedMortaja/fabric-samples/compare/patch-1...WaleedMortaja:PEER_CONN_PARMS)

**Comment by lehors** [comment]
Very nice! Well done.

**Comment by lehors** [comment]
Based on my tests, networkDown in network.sh also needs quotes around the use of $(pwd) or it fails to fully clean things up and spits out format errors:

```docker: invalid reference format.
See 'docker run --help'.
docker: invalid reference format.
See 'docker run --help'.
docker: invalid reference format.
See 'docker run --help'.
docker: invalid reference format.
See 'docker run --help'.
docker: invalid reference format.
See 'docker run --help'.
docker: invalid reference format.
See 'docker run --help'.
```

I think once this is done we'll have a complete fix.

Well, if you could do me a favor and also remove the sed command at the end of parsePeerConnectionParameters:

```
  # remove leading space for output
  PEERS="$(echo -e "$PEERS" | sed -e 's/^[[:space:]]*//')"
```
All that is needed is not to add the leading whitespace in the first place by doing:

```
    ## Set peer addresses
    if [ -z "$PEER" ]
    then
	PEERS="$PEER"
    else
	PEERS="$PEERS $PEER"
    fi
```

 Although not incorrect, the use of sed in this case hurts my head. :-)

Thank you!

**Comment by WaleedMortaja** [comment]
> Based on my tests, [networkDown in network.sh](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/test-network/network.sh#L317) also needs quotes around the use of $(pwd) or it fails to fully clean things up and spits out format errors:
> 
> ```
> See 'docker run --help'.
> docker: invalid reference format.
> See 'docker run --help'.
> docker: invalid reference format.
> See 'docker run --help'.
> docker: invalid reference format.
> See 'docker run --help'.
> docker: invalid reference format.
> See 'docker run --help'.
> docker: invalid reference format.
> See 'docker run --help'.
> ```
> 
> I think once this is done we'll have a complete fix.
> 
> Well, if you could do me a favor and also remove t[he sed command at the end of parsePeerConnectionParameters](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/test-network/scripts/envVar.sh#L92):
> 
> ```
>   # remove leading space for output
>   PEERS="$(echo -e "$PEERS" | sed -e 's/^[[:space:]]*//')"
> ```
> 
> All that is needed is not to add the leading whitespace in [the first place](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/test-network/scripts/envVar.sh#L83) by doing:
> 
> ```
>     ## Set peer addresses
>     if [ -z "$PEER" ]
>     then
> 	PEERS="$PEER"
>     else
> 	PEERS="$PEERS $PEER"
>     fi
> ```
> 
> Although not incorrect, the use of sed in this case hurts my head. :-)
> 
> Thank you!

I will do that.
By the way, do you have any recommendation about commits? make the last part a separe commit? or make it with the coming same commit? or maybe squashing all the commits in this pull request or something?

**Comment by WaleedMortaja** [comment]
> /azp run

It is an old comment, but I did not get what does it mean?

**Comment by WaleedMortaja** [comment]
I did not understand these lies ([here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/digibank/digibank.sh#L29) and [here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/magnetocorp/magnetocorp.sh#L29)). Do they need to be handled for whitespaces?

**Comment by WaleedMortaja** [comment]
>     ## Set peer addresses
>     if [ -z "$PEER" ]

I assume you want
` if [ -z "$PEERS" ]`

**Comment by lehors** [comment]
> I will do that.
> By the way, do you have any recommendation about commits? make the last part a separe commit? or make it with the coming same commit? or maybe squashing all the commits in this pull request or something?

It doesn't really matter they will get squashed before merging.

**Comment by lehors** [comment]
> > /azp run
> 
> It is an old comment, but I did not get what does it mean?

This is actually a command to get the CI pipeline triggered again. This is useful when the results show a failure that is believed to be accidental.

**Comment by lehors** [comment]
> > ```
> > ## Set peer addresses
> > if [ -z "$PEER" ]
> > ```
> 
> I assume you want
> ` if [ -z "$PEERS" ]`

You're correct.

**Comment by lehors** [comment]
> I did not understand these lies ([here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/digibank/digibank.sh#L29) and [here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/magnetocorp/magnetocorp.sh#L29)). Do they need to be handled for whitespaces?

These commands trim any leading spaces. It's actually a much more efficient way than the sed command I asked you to delete. But I actually don't see why this would matter so I think it is safe to delete those lines. If needed this should be done in parsePeerConnectionParameters instead, like it's done for $PEERS.

**Comment by WaleedMortaja** [comment]
> > I did not understand these lies ([here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/digibank/digibank.sh#L29) and [here](https://github.com/hyperledger/fabric-samples/blob/6fd327a347b8b0175ad83817ad4305eef5dfed29/commercial-paper/organization/magnetocorp/magnetocorp.sh#L29)). Do they need to be handled for whitespaces?
> 
> These commands trim any leading spaces. It's actually a much more efficient way than the sed command I asked you to delete. But I actually don't see why this would matter so I think it is safe to delete those lines. If needed this should be done in parsePeerConnectionParameters instead, like it's done for $PEERS.

Hello @lehors 
I have already done that, and thank you for all the information in the comments.
Are there any edits required before merging?

**Review by lehors — DISMISSED** [review]
Everything looks good and a test I ran went through all right.
Thanks for all your work on this!

**Review by lehors — APPROVED** [review]
I resolved a conflict with master. LGTM.

**Comment by WaleedMortaja** [comment]
> Everything looks good and a test I ran went through all right.
> Thanks for all your work on this!

You are welcome!
It was my honor to contribute to the project.

**Comment by WaleedMortaja** [comment]
The documents also need handle spaces when absolute paths of files are passed to commands. As in [here](https://github.com/hyperledger/fabric/blame/5df63f14258f8d9ed487d47429fbff142973962b/docs/source/deploy_chaincode.md#L392) for example:
`peer lifecycle chaincode approveformyorg -o localhost:7050 --ordererTLSHostnameOverride orderer.example.com --channelID mychannel --name basic --version 1.0 --package-id $CC_PACKAGE_ID --sequence 1 --tls --cafile ${PWD}/organizations/ordererOrganizations/example.com/orderers/orderer.example.com/msp/tlscacerts/tlsca.example.com-cert.pem`

I am not sure I will have time to fix it there my self there but I think It would be easier to fix using the edits exiting in this pull request. I will make an issue in the documents repo for the current time.

**Comment by lehors** [comment]
it's all right. I'll take a crack at fixing the doc. Thanks.

**Comment by lehors** [comment]
Done: https://github.com/hyperledger/fabric/pull/2298

**Comment by WaleedMortaja** [comment]
@Mergifyio backport release-2.2

**Comment by WaleedMortaja** [comment]
@Mergifyio refresh
